### PR TITLE
Introduced the "extract" function to all the parsers, with relative tests

### PR DESCRIPTION
**Pull request by DrDomenicoMarson**
Ok, I hope not to have broken anything with this PR :-)
I added the `extract(file, T)` function to all the existing parsers, with a relative test for each parser.

The only parser that was heavily hit by this PR is the AMBER parser, as was mentioned in issue #222 .
Here I had to make different adjustments, on my machine all tests are passing through. 
I took the opportunity to make also some small, cosmetic, changes to the code in the amber parser (just a couple of them, like removing the unnecessary 'class SomeThing(**Object**').

PS: brace yourself, I now have a copy of Amber2022, and it seems the way dHdl/MBAR data is written has changed, so the parser will need an urgent update to handle AMBER <2022 | >=2022)

### CODE REVIEW

File: src/alchemlyb/parsing/amber.py
Comment (xiki-tempula) [inline src/alchemlyb/parsing/amber.py:251]: Please don't change this as this would break sphinx.
Just say return a dictionary. Might also be good to check if the doc builds correctly.

```
Returns
-------
Dict
    A dictionary with keys of 'u_nk', which is a pandas DataFrame ....
```

File: src/alchemlyb/parsing/gmx.py
Comment (xiki-tempula) [inline src/alchemlyb/parsing/gmx.py:246]: Add the decorator

File: src/alchemlyb/parsing/gomc.py
Comment (xiki-tempula) [inline src/alchemlyb/parsing/gomc.py:152]: Add the decorator

File: src/alchemlyb/parsing/namd.py
Comment (xiki-tempula) [inline src/alchemlyb/parsing/namd.py:310]: Add the decorator

File: src/alchemlyb/tests/parsing/test_amber.py
Comment (xiki-tempula) [inline src/alchemlyb/tests/parsing/test_amber.py:56]: I like these way of just put the numbers there.

File: src/alchemlyb/tests/parsing/test_gomc.py
Comment (xiki-tempula) [inline src/alchemlyb/tests/parsing/test_gomc.py:39]: Just do this on a single file should be fine. No need to go through all the files.

**Review by xiki-tempula — CHANGES_REQUESTED** [review]
These are great changes. This is an initial review. Do you mind bumping the coverage a bit? Thanks,

**Comment by DrDomenicoMarson** [comment]
> These are great changes. This is an initial review. Do you mind bumping the coverage a bit? Thanks,

This is unfortunately impossible right now. I made a detailed check, the lines that are not covered by the tests are unaffected/not created by my modifications (so the coverage is the same as before, just the total number of lines is reduced as I removed an entire function in the amber parser).
I would like not to touch the other parsers, as I'm not an expert in doing TI/FEP with other software. In issue #241 I pointed out a place where we could increase coverage.

In the amber parser, I could easily reach a coverage of 99%, but the issues in alchemtest have to be addressed before, as I need 3 "properly" created files. Currently, in the amber parser, we have 4 lines that are not covered by the tests (and were not covered also before this PR), 3 of which are easily addressable with the 3 new files, while for the 4th I have to go deeper in understanding a function that was already there and that I don't currently understand 100% :-).

File: src/alchemlyb/parsing/amber.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/amber.py]: That's great, I was worried I had to go through an ISSUE/PR cycle just for that :-)

File: src/alchemlyb/parsing/amber.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/amber.py:149]: That was a "leftover" from a previous PR. Originally we extracted the components that are reported if `ntave` is used in the amber input, but we never used them. These components were saved in the `component_gradient` attribute of `FEData`. With a previous PR #230 I removed the component extraction, but I forgot to remove the attribute from the `__slots__` :-)

File: src/alchemlyb/parsing/amber.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/amber.py:236]: You are right, force of (new) habit :-)

File: src/alchemlyb/parsing/amber.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/amber.py:247]: I'm sorry, I didn't know about that :-)

Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/amber.py]: I edited the docstring for every new `extract` functions, and added to the docs, now the `extract` function appears correctly when the docs are built.

File: src/alchemlyb/parsing/gmx.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/gmx.py:246]: The decorator shouldn't be needed, as the `extract` function just returns a `dict` that is built calling the original `extract_*` functions, which have the correct decorator applied to them.
The only `extract` function that needs the new decorator is the AMBER one, as here it's the other way around, it's the `extract_dHdl/extract_u_nk` that calls the `extract` function.

File: src/alchemlyb/parsing/gomc.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/gomc.py:152]: Same as for gmx

File: src/alchemlyb/parsing/namd.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/namd.py:310]: same as for gmx

File: src/alchemlyb/tests/parsing/test_amber.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/tests/parsing/test_amber.py:56]: Yeah, I used the other way before just because other already existing tests were made like that, but I think it's cleaner like that...

File: src/alchemlyb/tests/parsing/test_gomc.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/tests/parsing/test_gomc.py:39]: You are right :-)

**Review by xiki-tempula — COMMENTED** [review]
The current one looks reasonable to me. Just some comments.

> This is unfortunately impossible right now. I made a detailed check, the lines that are not covered by the tests are unaffected/not created by my modifications (so the coverage is the same as before, just the total number of lines is reduced as I removed an entire function in the amber parser).

I do understand that some of the code are not being covered by the previous tests, do you mind adding `# pragma: no cover` to bump the coverage?

**Comment by DrDomenicoMarson** [comment]
`# pragma: no cover` added to AMBER parser, it feels like cheating though :-)

**Comment by xiki-tempula** [comment]
@DrDomenicoMarson Obviously, it would be best that all lines are covered. 
Obviously, some lines are more difficult to cover and offer little value to cover, so it is fine that those are not being covered. 
What we don't want is a surprise where we thought that some lines of code are being covered but they are not, thus, `# pragma: no cover` will tell you that this line is not covered.

**Comment by xiki-tempula** [comment]
Have you checked that sphinx builds? As RTD is failing?

**Comment by DrDomenicoMarson** [comment]
> Have you checked that sphinx builds? As RTD is failing?

I just re-checked, in my computer if I go inside alchemtest/docs and run `make html` it creates a perfectly fine HTML file, with the `extract` methods reported. 
sphinx gives me some warning, but concerning `alchemlyb.workflows.*` stuff, so I supposed those were already there!

**Comment by DrDomenicoMarson** [comment]
Ok, digging deeper the problem was unrelated to this PR.
sphinx was complaining about the first lines in `postprocessors/units.py'
my sphinx (which is v4.3.2 vs the v5 here) didn't complain about that.
For now I tried just commenting the offending lines (maybe a better solution exists?)

File: src/alchemlyb/tests/parsing/test_namd.py
Comment (xiki-tempula) [inline src/alchemlyb/tests/parsing/test_namd.py:311]: Why don't assert None here?

File: src/alchemlyb/parsing/namd.py
Comment (xiki-tempula) [inline src/alchemlyb/parsing/namd.py:343]: .. versionadded:: 1.0.0

File: src/alchemlyb/parsing/gomc.py
Comment (xiki-tempula) [inline src/alchemlyb/parsing/gomc.py:176]: .. versionadded:: 1.0.0

File: src/alchemlyb/parsing/gmx.py
Comment (xiki-tempula) [inline src/alchemlyb/parsing/gmx.py:270]: .. versionadded:: 1.0.0

**Comment by DrDomenicoMarson** [comment]
Ok, it's a good thing this happened for us, as I fixed also a couple of other things in the docs regarding `extract`, but the problem is not on us, it is from the sphinx side, as [https://github.com/readthedocs/sphinx_rtd_theme/issues/1343]

File: src/alchemlyb/parsing/amber.py
Comment (xiki-tempula) [inline src/alchemlyb/parsing/amber.py:252]: .. versionadded:: 1.0.0

**Review by xiki-tempula — CHANGES_REQUESTED** [review]
I'm happy with the content but we cannot merge it when the RTD fails.

**Comment by xiki-tempula** [comment]
Ok, probably give it a day or two and see if it got fixed by sphinx.

File: src/alchemlyb/parsing/namd.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/namd.py:343]: I don't understand this sorry, do I have to do anything?

Comment (xiki-tempula) [reply src/alchemlyb/parsing/namd.py:343]: If you check https://alchemlyb.readthedocs.io/en/latest/preprocessing/alchemlyb.preprocessing.subsampling.html#alchemlyb.preprocessing.subsampling.decorrelate_u_nk
You could see `New in version 0.6.0.`
You would need .. versionadded:: 1.0.0 as well.

**Comment by orbeckst** [comment]
> PS: brace yourself, I now have a copy of Amber2022, and it seems the way dHdl/MBAR data is written has changed, so the parser will need an urgent update to handle AMBER <2022 | >=2022)

Argh. Sigh. Thanks for raising #242 , that's the right approach!

**Comment by orbeckst** [comment]
Regarding coverage: Are you saying that if we merged https://github.com/alchemistry/alchemtest/pull/71 then you could increase the test coverage here and avoid no-cover pragmas?

**Comment by DrDomenicoMarson** [comment]
@orbeckst  Yeah, I can cover all lines of parsers/amber.py except 1 line, but maybe it's better not in this PR, I'd prefer to merge this with the `pragma`, then merge https://github.com/alchemistry/alchemtest/pull/71 and I'll submit a PR that removes `pragma`.

@xiki-tempula they fixed the bug in sphinx, so now all checks are passed here!

**Review by xiki-tempula — CHANGES_REQUESTED** [review]
Do you mind add the versionadded and this PR is ready to go.

File: src/alchemlyb/parsing/gmx.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/gmx.py:272]: Great, done!

File: src/alchemlyb/parsing/amber.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/amber.py:252]: Great, done!

File: src/alchemlyb/parsing/gomc.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/gomc.py:178]: Great, done!

File: src/alchemlyb/parsing/namd.py
Comment (DrDomenicoMarson) [reply src/alchemlyb/parsing/namd.py:345]: Great, done!

File: src/alchemlyb/parsing/amber.py
Comment (xiki-tempula) [reply src/alchemlyb/parsing/amber.py:252]: Thanks. Have you checked that the doc looks good? I was under the impression that the .. versionadded:: 1.0.0 need to blank line before it but I might be wrong.

**Review by xiki-tempula — APPROVED** [review]
LGTM will merge if all tests passes.

**Comment by DrDomenicoMarson** [comment]
@xiki-tempula what do you think about the `#Note` in the namd.py parser? Should we keep it like that, with just returning `u_nk`, or should we also add the item `dHdl: None` to be consistent with the other parsers?

**Comment by xiki-tempula** [comment]
@DrDomenicoMarson I think at the moment both of them are good. I mean it is either `if 'dhdl' in extract()` or `if extract()['dhdl'] is None`. Both of them are quite intuitive and both of them are ok with me.

**Comment by DrDomenicoMarson** [comment]
Ok, so we can just leave it like it's now, I remove the commented line in the test then, before the final merging!

**Pull request merged**
